### PR TITLE
wa/prepmod normalizer

### LINKED ISSUE #124: normalize prepmod from WA
[learn our pipeline: normalize](https://github.com/CAVaccineInventory/vaccine-feed-ingest/wiki/Runner-pipeline-stages#normalize)

Blocked on #123 

Normalize existing [`.ndjson`](http://ndjson.org/) records into [the Vaccinate The States data schema](https://github.com/CAVaccineInventory/vaccine-feed-ingest/blob/main/vaccine_feed_ingest/schema/schema.py).

Read all files in the directory passed as the second argument (`sys.argv[2]`), transform each record into the normalized schema (filling in as much detail as possible), then output them to new files in the directory passed as the first argument (`sys.argv[1]`).

Check the wiki to learn more about the purpose of the normalize stage and how to get set up for development!

### Tips

1. Fetch and parse data for this site before you start developing:
    ```sh
    poetry run vaccine-feed-ingest fetch wa/prepmod
    poetry run vaccine-feed-ingest parse wa/prepmod
    ```

1. While working on your code, run it at any point:
    ```sh
    poetry run vaccine-feed-ingest normalize wa/prepmod
    ```

### Example
[Parsed files for `ak/arcgis`](https://github.com/CAVaccineInventory/vaccine-feed-ingest-results/tree/main/ak/arcgis/parsed) are converted to [normalized files](https://github.com/CAVaccineInventory/vaccine-feed-ingest-results/tree/main/ak/arcgis/normalized)

### PR DESCRIPTION
**Pull request by tmb**
Normalize Prepmod for WA

| Key Details |
|-|
| Resolves #124  |
State: WA |
Site: prepmod |

## Notes
Some clinics have links/phone numbers in the notes field, but they aren't reliably the contact info for that clinic (they can be emergency numbers / links to medical record sites / etc), so I kept that info in the notes and didn't parse out separately.

## Data sample
```json
{"id": "wa_prepmod:2731", "name": "Rebel Med NW - COVID Vaccine Clinic", "address": {"street1": "5401 Leary Ave NW", "street2": null, "city": "Seattle", "state": "WA", "zip": "98107"}, "location": null, "contact": [{"contact_type": "booking", "phone": null, "website": "https://prepmod.doh.wa.gov/client/registration?clinic_id=2731", "email": null, "other": null}], "languages": null, "opening_dates": [{"opens": "2021-04-30", "closes": "2021-04-30"}], "opening_hours": [{"day": "Friday", "open": "09:00", "closes": "17:00"}], "availability": {"drop_in": null, "appointments": true}, "inventory": [{"vaccine": "moderna", "supply_level": null}], "access": null, "parent_organization": null, "links": null, "notes": ["***Currently only Moderna COVID First Dose Vaccines Available***\n***IMPORTANT PLEASE READ ABOVE*** You must provide insurance (photo ID if you do not have insurance) by one of the 2 methods below before you check-in for vaccine. Upload options 1. Prepmod at time of sigh up 2. prior to appointment by contacting info@rebelmednw.com\n\nWe are administering the first dose of the Moderna vaccine. Moderna is currently approved for all adults 18 years and older. Please bring a face covering that covers the mouth and nose. Everyone is required to stay 15-30 minutes under observation after receiving the vaccine.\n\nPlease contact 206-297-6013, and as to speak to a provider on call. Call 911 if life-threatening emergency, and you are in need of immediate help.", "If you are signing up for a second dose, you must get the same vaccine brand as your first dose."], "active": null, "source": {"source": "wa_prepmod", "id": "Rebel Med NW - COVID Vaccine Clinic", "fetched_from_uri": "https://prepmod.doh.wa.gov/clinic/search", "fetched_at": "2021-04-26T03:41:43.854225", "published_at": null, "data": {"name": "Rebel Med NW - COVID Vaccine Clinic", "date": "04/30/2021", "address": "5401 Leary Ave NW, Seattle WA, 98107", "vaccines": "Moderna COVID-19 Vaccine", "ages": "Adults, Seniors", "info": "***Currently only Moderna COVID First Dose Vaccines Available***\n***IMPORTANT PLEASE READ ABOVE*** You must provide insurance (photo ID if you do not have insurance) by one of the 2 methods below before you check-in for vaccine. Upload options 1. Prepmod at time of sigh up 2. prior to appointment by contacting info@rebelmednw.com\n\nWe are administering the first dose of the Moderna vaccine. Moderna is currently approved for all adults 18 years and older. Please bring a face covering that covers the mouth and nose. Everyone is required to stay 15-30 minutes under observation after receiving the vaccine.\n\nPlease contact 206-297-6013, and as to speak to a provider on call. Call 911 if life-threatening emergency, and you are in need of immediate help.", "hours": "09:00 am - 05:00 pm", "available": "14", "special": "If you are signing up for a second dose, you must get the same vaccine brand as your first dose.", "clinic_id": "2731"}}}
{"id": "wa_prepmod:1997", "name": "Cle Elum Community Clinic", "address": {"street1": "2690 WA-903", "street2": null, "city": "Cle Elum", "state": "WA", "zip": "98922"}, "location": null, "contact": [{"contact_type": "booking", "phone": null, "website": "https://prepmod.doh.wa.gov/client/registration?clinic_id=1997", "email": null, "other": null}], "languages": null, "opening_dates": [{"opens": "2021-05-01", "closes": "2021-05-01"}], "opening_hours": [{"day": "Saturday", "open": "12:00", "closes": "18:00"}], "availability": {"drop_in": null, "appointments": true}, "inventory": null, "access": null, "parent_organization": null, "links": null, "notes": ["", "If you are signing up for a second dose, you must get the same vaccine brand as your first dose."], "active": null, "source": {"source": "wa_prepmod", "id": "Cle Elum Community Clinic", "fetched_from_uri": "https://prepmod.doh.wa.gov/clinic/search", "fetched_at": "2021-04-26T03:41:43.854225", "published_at": null, "data": {"name": "Cle Elum Community Clinic", "date": "05/01/2021", "address": "2690 WA-903, Cle Elum WA, 98922", "vaccines": "Contact Clinic Site", "ages": "Adults", "info": "", "hours": "12:00 pm - 06:00 pm", "available": "414", "special": "If you are signing up for a second dose, you must get the same vaccine brand as your first dose.", "clinic_id": "1997"}}}
{"id": "wa_prepmod:2015", "name": "SRHD- Fire Station - Emergency Management Services", "address": {"street1": "929 South Garfield Road", "street2": null, "city": "Airway Heights", "state": "WA", "zip": "99001"}, "location": null, "contact": [{"contact_type": "booking", "phone": null, "website": "https://prepmod.doh.wa.gov/client/registration?clinic_id=2015", "email": null, "other": null}], "languages": null, "opening_dates": [{"opens": "2021-05-01", "closes": "2021-05-01"}], "opening_hours": [{"day": "Saturday", "open": "08:00", "closes": "12:00"}], "availability": {"drop_in": null, "appointments": true}, "inventory": [{"vaccine": "pfizer_biontech", "supply_level": null}], "access": null, "parent_organization": null, "links": null, "notes": ["Vaccines by appointment only.  No Walk-ins", "If you are signing up for a second dose, you must get the same vaccine brand as your first dose."], "active": null, "source": {"source": "wa_prepmod", "id": "SRHD- Fire Station - Emergency Management Services", "fetched_from_uri": "https://prepmod.doh.wa.gov/clinic/search", "fetched_at": "2021-04-26T03:41:43.854225", "published_at": null, "data": {"name": "SRHD- Fire Station - Emergency Management Services", "date": "05/01/2021", "address": "929 South Garfield Road, Airway Heights WA, 99001", "vaccines": "Pfizer-BioNTech COVID-19 Vaccine", "ages": "Adults, Seniors, Other", "info": "Vaccines by appointment only.  No Walk-ins", "hours": "08:00 am - 12:00 pm", "available": "28", "special": "If you are signing up for a second dose, you must get the same vaccine brand as your first dose.", "clinic_id": "2015"}}}
{"id": "wa_prepmod:2583", "name": "Camano Island Fire and Rescue- COVID Vaccination Site", "address": {"street1": "606 Arrowhead Road", "street2": null, "city": "Camano", "state": "WA", "zip": "98282"}, "location": null, "contact": [{"contact_type": "booking", "phone": null, "website": "https://prepmod.doh.wa.gov/client/registration?clinic_id=2583", "email": null, "other": null}], "languages": null, "opening_dates": [{"opens": "2021-05-01", "closes": "2021-05-01"}], "opening_hours": [{"day": "Saturday", "open": "09:30", "closes": "16:30"}], "availability": {"drop_in": null, "appointments": true}, "inventory": [{"vaccine": "moderna", "supply_level": null}], "access": null, "parent_organization": null, "links": null, "notes": ["FIRST DOSE APPOINTMENTS ONLY PLEASE. Other appointment requests will be cancelled", "If you are signing up for a second dose, you must get the same vaccine brand as your first dose."], "active": null, "source": {"source": "wa_prepmod", "id": "Camano Island Fire and Rescue- COVID Vaccination Site", "fetched_from_uri": "https://prepmod.doh.wa.gov/clinic/search", "fetched_at": "2021-04-26T03:41:43.854225", "published_at": null, "data": {"name": "Camano Island Fire and Rescue- COVID Vaccination Site", "date": "05/01/2021", "address": "606 Arrowhead Road, Camano WA, 98282", "vaccines": "Moderna COVID-19 Vaccine", "ages": "Adults", "info": "FIRST DOSE APPOINTMENTS ONLY PLEASE. Other appointment requests will be cancelled", "hours": "09:30 am - 04:30 pm", "available": "512", "special": "If you are signing up for a second dose, you must get the same vaccine brand as your first dose.", "clinic_id": "2583"}}}
```

## Before Opening a PR
- [x] I tested this using the CLI (e.g., `poetry run vaccine-feed-ingest <state>/<site>`)
- [x] I ran auto-formatting:
  - [x] `poetry run black vaccine_feed_ingest`
  - [x] `poetry run isort vaccine_feed_ingest`

### CODE REVIEW

**Comment by tmb** [comment]
just had an idea for how to put the clinic sign-up link. don't merge yet!

**Comment by tmb** [comment]
sign-up link added! all set for review again.

File: vaccine_feed_ingest/runners/wa/prepmod/normalize.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    return schema.Source(
+        source="wa_prepmod",
+        id=site["name"],
+        fetched_from_uri="https://prepmod.doh.wa.gov/clinic/search",
+        fetched_at=timestamp,
+        data=site,
+    )
+
+
+def _get_inventory(site: dict) -> List[schema.Vaccine]:
```
Comment (eliblock) [inline vaccine_feed_ingest/runners/wa/prepmod/normalize.py:24]: I wonder if we could offer something like this as a shared helper which folks could use in several normalizers

**Review by eliblock — APPROVED** [review]
Thank you for adding this - overall it looks really good!

I'm a bit concerned about the flood of near-duplicates (e.g., `Island Drug - COVID Vaccine Site` has multiple single-day clinics on the same day with different times at the same place, and overall has a series of single-day clinics which will be interesting), but this is a feature of the source data, not an issue with normalization. CC: @bryanculbertson for awareness.

File: vaccine_feed_ingest/runners/wa/prepmod/normalize.py (hunk 1 above)
Comment (tmb) [reply vaccine_feed_ingest/runners/wa/prepmod/normalize.py:24]: yeah, agree.

**Pull request merged**